design/src/rfcs/rfc_0020_endpoints_20.md

```diff
@@ -0,0 +1,597 @@
+                            self
+                        }
+                        """,
+                        param.type
+                    )
+                }
+            }
+            ServiceConfig.BuilderBuild -> writable {
+                contextParams.forEach { param ->
+                    rust("${param.name}: self.${param.name},")
+                }
+            }
+            else -> emptySection
+        }
+    }
+}
+```
+
+</details>
+
+### Creating `Params`
+
+`Params` will be created and utilized in smithy-native code generation.
+
+`make_operation()` needs to load the parameters from several configuration sources. These sources have a priority order.
+To handle this priority order, we will load from all sources in reverse priority order, with lower priority sources
+overriding higher priority ones.
+
+<details>
+<summary>Implementation of operation decorator</summary>
+
+```kotlin
+class EndpointParamsDecorator(
+    private val ctx: ClientCodegenContext,
+    private val operationShape: OperationShape,
+) : OperationCustomization() {
+    val idx = ContextIndex.of(ctx.model)
+    private val ruleset = EndpointRuleset.fromNode(ctx.serviceShape.expectTrait<EndpointRuleSetTrait>().ruleSet)
+
+    override fun section(section: OperationSection): Writable {
+        return when (section) {
+            is OperationSection.MutateInput -> writable {
+                rustTemplate(
+                    """
+                    let params = #{Params}::builder()
+                        #{builder:W}.expect("invalid endpoint");
+                    """,
+                    "Params" to EndpointParamsGenerator(ruleset).paramsStruct(),
+                    "builder" to builderFields(section)
+                )
+            }
+            is OperationSection.MutateRequest -> writable {
+                rust("// ${section.request}.properties_mut().insert(params);")
+            }
+            else -> emptySection
+        }
+    }
+
+    private fun builderFields(section: OperationSection.MutateInput) = writable {
+        val memberParams = idx.getContextParams(operationShape)
+        val builtInParams = ruleset.parameters.toList().filter { it.isBuiltIn }
+        // first load builtins and their defaults
+        builtInParams.forEach { param ->
+            val defaultProviders = section.endpointCustomizations.mapNotNull { it.defaultFor(param, section.config) }
+            if (defaultProviders.size > 1) {
+                error("Multiple providers provided a value for the builtin $param")
+            }
+            defaultProviders.firstOrNull()?.also { defaultValue ->
+                rust(".set_${param.name.rustName()}(#W)", defaultValue)
+            }
+        }
+        // these can be overridden with client context params
+        idx.getClientContextParams(ctx.serviceShape).forEach { (name, _param) ->
+            rust(".set_${name.toSnakeCase()}(${section.config}.${name.toSnakeCase()}.as_ref())")
```

Comment:
is `rustName` the same as `toSnakeCase`? Can we just use one or the other?

Reply:
`rustName()` is an extension function on `Identifier` to do snakeCasing & escaping.